src/ui/shared/UrlContext.ts

```diff
@@ -0,0 +1,97 @@
+export type AppMode = 'onboarding' | 'wallet';
+export type WindowType = 'popup' | 'tab' | 'dialog';
+export type WindowLayout = 'column' | 'page';
+
+export enum UrlContextParam {
+  windowType = 'windowType',
+  windowLayout = 'windowLayout',
+  appMode = 'appMode',
+}
+
+function getUrlContextParam<T>(param: UrlContextParam, defaultValue?: T) {
+  const url = new URL(window.location.href);
+  return (url.searchParams.get(param) as T) || defaultValue;
+}
+
+export interface UrlContextParams {
+  appMode?: AppMode;
+  windowType?: WindowType;
+  windowLayout?: WindowLayout;
+}
+
+class UrlContext {
+  protected params: UrlContextParams;
+
+  constructor(params: UrlContextParams) {
+    this.params = params;
+  }
+
+  set(searchParams: URLSearchParams, params: UrlContextParams) {
```

Comment:
The only way to set the URL context parameters is by using this `set` function. It also keeps our global params object (see below) in sync with the window's `URLSearchParams`